File: apps/worker/src/render-docx.ts
import type { DocumentValue } from "@aqshara/documents";
import {
  Document,
  HeadingLevel,
  Packer,
  Paragraph,
  TextRun,
} from "docx";

function headingLevel(level: 1 | 2 | 3): (typeof HeadingLevel)[keyof typeof HeadingLevel] {
  if (level === 1) return HeadingLevel.HEADING_1;
  if (level === 2) return HeadingLevel.HEADING_2;
  return HeadingLevel.HEADING_3;
}

function blockText(block: DocumentValue[number]): string {
  if (block.type === "bullet-list") {
    return block.children
      .map((li) => li.children.map((t) => t.text).join(""))
      .join("\n");
  }
  return block.children.map((c) => c.text).join("");
}

export async function renderDocumentValueToDocxBuffer(input: {
  title: string;
  value: DocumentValue;
}): Promise<Buffer> {
  const paragraphs: Paragraph[] = [];

  if (input.title.trim()) {
    paragraphs.push(
      new Paragraph({
        children: [new TextRun({ text: input.title.trim() })],
        heading: HeadingLevel.TITLE,
      }),
    );
  }

  for (const block of input.value) {
    if (block.type === "heading") {
      paragraphs.push(
        new Paragraph({
          children: [new TextRun({ text: blockText(block) })],
          heading: headingLevel(block.level),
        }),
      );
    } else if (block.type === "paragraph") {
      paragraphs.push(
        new Paragraph({
          children: [new TextRun({ text: blockText(block) })],
        }),
      );
    } else if (block.type === "bullet-list") {
      for (const li of block.children) {
        const text = li.children.map((t) => t.text).join("");
        paragraphs.push(
          new Paragraph({
            children: [new TextRun({ text })],
            bullet: { level: 0 },
          }),
        );
      }
    }
  }

  const doc = new Document({
    sections: [
      {
        children: paragraphs,
      },
    ],
  });

  return Buffer.from(await Packer.toBuffer(doc));
}

File: apps/worker/src/worker/jobs/render_docx.py
from __future__ import annotations

from docx import Document


def render_document_value_to_docx_bytes(*, title: str, value: list[dict[str, object]]) -> bytes:
    document = Document()
    if title.strip():
        document.add_heading(title.strip(), level=0)

    for block in value:
        block_type = block["type"]
        if block_type == "heading":
            level = int(block.get("level", 1))
            text = "".join(child["text"] for child in block["children"])
            document.add_heading(text, level=min(level, 3))
        elif block_type == "paragraph":
            text = "".join(child["text"] for child in block["children"])
            document.add_paragraph(text)
        elif block_type == "bullet-list":
            for item in block["children"]:
                text = "".join(child["text"] for child in item["children"])
                document.add_paragraph(text, style="List Bullet")

    buffer = b""
    from io import BytesIO

    io_buffer = BytesIO()
    document.save(io_buffer)
    buffer = io_buffer.getvalue()
    return buffer
